src/main/java/frc/robot/subsystems/LimeVision/LimeLightSub.java

```diff
@@ -117,4 +212,111 @@ public boolean checkPipe() {
   public double getError() {
     return PIDVision.calculate(getTx());
   }
+
+  // calculates x-distance error
+  public double getErrorFromMegaTagX() {
+    return PIDVision.calculate(getPoseX());
+  }
+
+  // calculates y-distance error
+  public double getErrorFromMegaTagY() {
+    return PIDVisionY.calculate(getPoseY());
+  }
+
+  // calculates y-distance error
+  public double getThetaErrorFromMegaTag() {
+    return PIDVisionTheta.calculate(getThetaZ());
+  }
+
+  public void setSetpointX(double point) {
+    PIDVision.setSetpoint(point);
+  }
+
+  public void setSetpointY(double point) {
+    PIDVisionY.setSetpoint(point);
+  }
+
+  public void setSetpointTheta(double point) {
+    PIDVisionTheta.setSetpoint(point);
+  }
+
+// megatag tx and ty, units in meters
+  public Pose2d getPose2D() {
+    return new Pose2d(Math.abs(getBotPose()[0]), Math.abs(getBotPose()[1]), Rotation2d.fromDegrees(getBotPose()[5]));
+  }
+
+  public double[] getBlueBotPose(){
+    double[] blueBotPose = this.blueBotPose.getDoubleArray(new double[11]);
+    double[] colorBotPose = new double[12];
+
+    colorBotPose[0] = blueBotPose[0];
+    colorBotPose[1] = blueBotPose[1];
+    colorBotPose[2] = blueBotPose[2];
+    colorBotPose[3] = blueBotPose[3];
+    colorBotPose[4] = blueBotPose[4];
+    colorBotPose[5] = blueBotPose[5];
+    colorBotPose[6] = blueBotPose[6];
+    colorBotPose[7] = blueBotPose[7];
+    colorBotPose[8] = blueBotPose[8];
+    colorBotPose[10] = blueBotPose[10];
+
+    if(hasTarget()){
+      colorBotPose[9] = blueBotPose[9];
+      colorBotPose[11] = getTID()+1;
+
+    }
+    
+
+    return colorBotPose; 
+  }
+
+  public double[] getTargetSpace(){
+    double[] targetSpace = this.targetSpace.getDoubleArray(new double[6]);
+    return targetSpace;
+  } 
+
+  public int getTID(){
+  
+    int tid = (int)this.tid.getDouble(-1)-1;
+    
+    return tid;
+    
+  }
+
+  public void setData(){
+    double[] temp = getBlueBotPose();
+    Pose2d currentPose = VisionContainer.drivebase.getPose();
```

Comment:
The drivebase should be an member of the LimeLightSub class, not accessed via a global variable